### PR TITLE
Use the same checks in all `grid_sampler` functions

### LINKED ISSUE #73187: grid_sampler_3d_backward aborts due to heap corruption
### 🐛 Describe the bug

Heap corruption in `grid_sampler_3d_backward`.

### Example to reproduce

```python
import torch

grad_output = torch.full((1, 1, 1, 4, 4,), 1, dtype=torch.float64, requires_grad=True)
input = torch.full((5, 5, 5, 5, 5,), 3.5e+35, dtype=torch.float64, requires_grad=True)
grid = torch.full((1, 1, 1, 4, 4,), 1, dtype=torch.float64, requires_grad=True)
interpolation_mode = 0
padding_mode = 0
align_corners = True
res = torch.grid_sampler_3d(input, grid, interpolation_mode, padding_mode, align_corners)
grad_out = torch.zeros_like(res)
torch.autograd.backward(res, grad_tensors=grad_out)
```

### Result
`sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed`
Program aborts due to heap corruption.

### Expected behavior
Heap does not get corrupted and program does not abort.

### Note
This bug was discovered using fuzz testing.

### Versions

PyTorch version: 1.9.0a0+git4d36e60
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: 11.0.1-2
CMake version: version 3.22.1
Libc version: glibc-2.31

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.10.0-11-amd64-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.22.2
[pip3] torch==1.9.0a0+git4d36e60
[conda] mkl                       2022.0.1           h06a4308_117
[conda] mkl-include               2022.0.1           h06a4308_117
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] torch                     1.9.0a0+git4d36e60           dev_0    <develop>

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74635

Fixes #73187.

Differential Revision: [D35284563](https://our.internmc.facebook.com/intern/diff/D35284563)

### CODE REVIEW

**Comment by nkaretnikov** [comment]
basically, the idea here is this: move the checks from the top-level `grid_sampler` function into a separate header/cpp file such that it can be used/linked with code on cpu and cuda, and use the said functions in all `grid_sampler*` kernels. added some basic tests to make sure things do not crash anymore when you pass in weird shapes

**Comment by nkaretnikov** [comment]
the changes to error messages in old files are due to me moving/splitting the checks for 4d/5d cases

**Review by albanD — COMMENTED** [review]
Thanks for the cleanup. it looks quite good. High level questions:
- Are the CI failures relevant?
- It would be much simpler if the utils functions were only in the header (no cpp) file. That would remove the need to update the build systemsss. Do you think that would be a better way to go here to make this change simpler?

**Comment by nkaretnikov** [comment]
@albanD i'll look at the ci failures tomorrow probably, but to answer your second question: i think the current way is actually the easiest. i also felt uncomfortable about adding new files, but i think you just have to do it for the following reasons:

- to support incremental (or whatever the proper term is) builds, that is, not depending on `native_functions.yaml`, as mentioned [here](https://github.com/pytorch/pytorch/pull/74635/files#diff-c0ea96a5e8d85793e72b2d8417b80d8f2a27fe8837e9b9ed37c43859c2809127R3):

```
// See NOTE: [Tensor vs. TensorBase]
// https://github.com/pytorch/pytorch/pull/66979
```

you need to avoid exposing `ATen/ATen.h` in the header and not using types/headers that may pull it in. which is why i use `TensorBase` instead of `Tensor` and `int64_t` instead of `GridSamplerInterpolation` in the header. i did have to move the function bodies out because the code needs both of these in the implementations. 

- otherwise, [here](https://github.com/pytorch/pytorch/pull/74635/files#diff-bb063818c22c14573e6d553e9ba46402cb866917d5d93c7ac0a0831fa4e62c49R1):

```
#define TORCH_ASSERT_NO_OPERATORS
#include <ATen/native/cuda/GridSampler.h>
#include <ATen/native/GridSamplerUtils.h>
```

i'd need to include the header before the assert (or it won't compile), and it would be undermining the work done by peter in the pr i linked above (#66979). so i simply don't see any other way to do it.

i did try the header-only thing as well, i even have a wip branch with that work already. i just need to double check that it actually works and i should be able to share it. but given the things i mentioned above, should i?

cc @peterbell10

File: aten/src/ATen/native/GridSamplerUtils.cpp
Comment (peterbell10) [inline aten/src/ATen/native/GridSamplerUtils.cpp:89]: The only part of this file that currently needs `Tensor` rather than `TensorBase` are these two function calls. Which themselves would work fine with `TensorBase`. So I would suggest adding a new overload `cudnn_is_acceptable(const TensorBase &)` and have the `const Tensor &` overload just call that version.

File: aten/src/ATen/CMakeLists.txt
Comment (peterbell10) [inline aten/src/ATen/CMakeLists.txt:120]: These CMake changes shouldn't be necessary. Just add `TORCH_API` to each function in `GridSamplerUtils.h`.

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/CMakeLists.txt
Comment (nkaretnikov) [reply aten/src/ATen/CMakeLists.txt:120]: i'll give it a try, but i did this because it failed during linking. iiuc, usually people put cuda code under `native/cuda` , but this one is one level above to allow sharing between cpu and cuda. it would work as is if it were a header-only change (you'd need to put the functions into an anonymous namespace tho to avoid multiple definition errors)

File: aten/src/ATen/native/GridSamplerUtils.cpp
Comment (nkaretnikov) [reply aten/src/ATen/native/GridSamplerUtils.cpp:89]: iiuc, not only these. if you change code like this:

```cpp
void check_grid_sampler_common(const TensorBase& input_, const TensorBase& grid_) {
  // auto input = Tensor(input_);
  // auto grid = Tensor(grid_);
  auto input = input_;
  auto grid = grid_;
...
```

and try to recompile, you'll run into issues with `operator<<`. should i add an overload for it as well?

on top of that, in `check_grid_sampler_3d` i compare against `GridSamplerInterpolation::Bicubic`. i tried really hard to avoid depending on `GridSamplerInterpolation` in the header because it's usually used where the original `GridSampler.h` header is included and it creates this dependency cycle on `native_functions.yaml` that's hard to untangle otherwise. in the header, i can use `int64_t` instead, which is fine, but in the implementation i need to know the value of `Bicubic` and i cannot just hardcode the int constant (otherwise, what's the point of having an enum in the first place?).

Comment (nkaretnikov) [reply aten/src/ATen/native/GridSamplerUtils.cpp:89]: again, i'm in favor of simplifying this and using the header-only approach (or anything else), but it's just not clear to me how to do it given the things i've mentioned.

**Comment by nkaretnikov** [comment]
pushed a small update to fix the linter issue and rocm builds. not sure if it fixes everything because i can only do cuda builds locally atm. let's see what ci thinks

File: aten/src/ATen/CMakeLists.txt
Comment (peterbell10) [reply aten/src/ATen/CMakeLists.txt:120]: That's exactly what `TORCH_API` does; it exports the symbol from `torch_cpu.so` so it can be used in `torch_cuda.so` and is linked correctly. There is no need to compile it twice.

File: aten/src/ATen/native/GridSamplerUtils.cpp
Comment (peterbell10) [reply aten/src/ATen/native/GridSamplerUtils.cpp:89]: >and try to recompile, you'll run into issues with operator<<. should i add an overload for it as well?

I see. I would printing `input.defined() ? "defined" : "undefined"` instead of printing `input` and `grid` themselves.

>i tried really hard to avoid depending on GridSamplerInterpolation in the header because it's usually used where the original `GridSampler.h` header is included and it creates this dependency cycle on `native_functions.yaml` that's hard to untangle otherwise.

I'm not sure what you mean. `GridSampler.h`only includes C++ standard library headers so I can't see any problems.

File: aten/src/ATen/native/GridSamplerUtils.cpp
Comment (peterbell10) [inline aten/src/ATen/native/GridSamplerUtils.cpp:4]:
```suggestion
#include <ATen/native/CanUse32BitIndexMath.h>
```

File: aten/src/ATen/native/GridSamplerUtils.cpp
Comment (nkaretnikov) [reply aten/src/ATen/native/GridSamplerUtils.cpp:89]: thanks for the tips! i think i've managed to do it. will push an update shortly

File: aten/src/ATen/CMakeLists.txt
Comment (nkaretnikov) [reply aten/src/ATen/CMakeLists.txt:120]: looks like this approach works, will push as part of the next update

File: aten/src/ATen/native/GridSamplerUtils.cpp
Comment (nkaretnikov) [reply aten/src/ATen/native/GridSamplerUtils.cpp:4]: will be part of the next update

**Comment by nkaretnikov** [comment]
pushed an update implementing the header-only approach -- no build system changes

**Comment by nkaretnikov** [comment]
i wonder if i should write more test cases such that i trigger every TORCH_CHECK. but it might be tricky, not sure it's worth it. thoughts?

**Review by peterbell10 — COMMENTED** [review]
`TensorBase` related stuff looks good to me. Didn't check the `grid_sampler` changes too closely but the tests are broken at the moment.

**Review by albanD — APPROVED** [review]
Looks good!
Thanks for taking the time to update the code to work well with TensorMeta!

**Comment by albanD** [comment]
The newly added test is failing in CI though

**Comment by nkaretnikov** [comment]
i'll take a look at the failures. will follow up later

**Comment by nkaretnikov** [comment]
couldn't make sense of the ci errors, so rebased and pushed. let's wait for ci again and see if this changes anything

**Comment by nkaretnikov** [comment]
will fix this rocm failure once the ci is done:

```
======================================================================
FAIL [0.023s]: test_invalid_shapes_grid_sampler_cuda (__main__.TestTorchDeviceTypeCUDA)
----------------------------------------------------------------------
RuntimeError: cudnn_grid_sampler_forward: ATen not compiled with cuDNN support

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1780, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "test_torch.py", line 1448, in test_invalid_shapes_grid_sampler
    torch.cudnn_grid_sampler(input, grid)
AssertionError: "expected grid and input to have same batch size" does not match "cudnn_grid_sampler_forward: ATen not compiled with cuDNN support"
```

**Comment by nkaretnikov** [comment]
does ci need to be updated?

https://github.com/pytorch/pytorch/runs/5761511611 

```
W: The repository 'https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64  Release' does not have a Release file.
W: The repository 'https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1604/x86_64  Release' does not have a Release file.
E: Failed to fetch https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/Packages  404  Not Found
E: Failed to fetch https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1604/x86_64/Packages  404  Not Found
```

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35284563).

**Comment by nkaretnikov** [comment]
okay, it's essential that that one test is fixed/working tho (before this can be merged) because it's the one i'm trying to repro (see the other pr in the stack for discussion on this)

**Comment by nkaretnikov** [comment]
i've just triggered a re-run because it looks like this is fixed on hud(?). not sure if i need to rebase again tho

**Comment by albanD** [comment]
The new CI run looks good. So I'm going to try to land as-is. Will let you know if this needs a rebase.

**Comment by albanD** [comment]
The cuda test failures were actually real and this broke master: https://github.com/pytorch/pytorch/runs/5789789694?check_suite_focus=true. Reverting.

**Pull request closed without merging**
